pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go

```diff
@@ -67,42 +68,67 @@ func (ba *BalancedAllocation) ScoreExtensions() framework.ScoreExtensions {
+	args, ok := baArgs.(*config.NodeResourcesBalancedAllocationArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type NodeResourcesBalancedAllocationArgs, got %T", baArgs)
+	}
+
+	if err := validation.ValidateNodeResourcesBalancedAllocationArgs(nil, args); err != nil {
+		return nil, err
+	}
+
+	resToWeightMap := make(resourceToWeightMap)
+	for _, resource := range args.Resources {
+		resToWeightMap[v1.ResourceName(resource.Name)] = 1
+	}
+
 	return &BalancedAllocation{
 		handle: h,
 		resourceAllocationScorer: resourceAllocationScorer{
 			Name:                BalancedAllocationName,
 			scorer:              balancedResourceScorer,
-			resourceToWeightMap: defaultRequestedRatioResources,
+			resourceToWeightMap: resToWeightMap,
 			enablePodOverhead:   fts.EnablePodOverhead,
 		},
 	}, nil
 }
 
-// todo: use resource weights in the scorer function
 func balancedResourceScorer(requested, allocable resourceToValueMap) int64 {
-	cpuFraction := fractionOfCapacity(requested[v1.ResourceCPU], allocable[v1.ResourceCPU])
-	memoryFraction := fractionOfCapacity(requested[v1.ResourceMemory], allocable[v1.ResourceMemory])
-	// fractions might be greater than 1 because pods with no requests get minimum
-	// values.
-	if cpuFraction > 1 {
-		cpuFraction = 1
-	}
-	if memoryFraction > 1 {
-		memoryFraction = 1
+	resourceToFractions := []float64{}
+	var totalFraction float64
+	for name, value := range requested {
+		// This could only happen if both requsted[resource] and allocable[resource] are zero, this case could be avoided by tainting
+		// the node so that the node with the scare resources could be untouched and leave more room for the pods which do need them.
+		if allocable[name] == 0 {
+			continue
+		}
+		fraction := float64(value) / float64(allocable[name])
+		if fraction > 1 {
+			fraction = 1
+		}
+		totalFraction += fraction
+		resourceToFractions = append(resourceToFractions, fraction)
 	}
 
-	// Upper and lower boundary of difference between cpuFraction and memoryFraction are -1 and 1
-	// respectively. Multiplying the absolute value of the difference by `MaxNodeScore` scales the value to
-	// 0-MaxNodeScore with 0 representing well balanced allocation and `MaxNodeScore` poorly balanced. Subtracting it from
-	// `MaxNodeScore` leads to the score which also scales from 0 to `MaxNodeScore` while `MaxNodeScore` representing well balanced.
-	diff := math.Abs(cpuFraction - memoryFraction)
-	return int64((1 - diff) * float64(framework.MaxNodeScore))
-}
-
-func fractionOfCapacity(requested, capacity int64) float64 {
-	if capacity == 0 {
-		return 1
+	std := 0.0
+
+	// For most cases, resource are limited to cpu and memory, the std could be simplified to std := (fraction1-fraction2)/2
+	// len(fractions) > 2: caculate std based on the well-known formula - root square of Σ((fraction(i)-mean)*(fraction(i)-mean))/len(fractions)
+	// Otherwise, set the std to zero is enough.
```

Comment:
not sure if this was discussed before, but what is the difference between the std approach and simply taking the difference between the max and min fraction as follows:

`(1-(max(fractions)-min(fractions)))*100`

the max-min differentiates better, but may be we don't want this plugin to give such a strong signal?

<google-sheets-html-origin><!--td {border: 1px solid #ccc;}br {mso-data-placement:same-cell;}-->

  | Node 1 | Node 2
-- | -- | --
resource 1 | 0.10 | 0.30
resource 2 | 0.30 | 0.30
resource 3 | 0.80 | 0.35
  |   |  
std | 63.94 | 97.11
max-min | 30.00 | 95.00

Reply:
yes, this have been discussed here, pls check out the discussion started from here: https://github.com/kubernetes/kubernetes/pull/101946#discussion_r631830234

Reply:
but seems like we have not considered this formula, 
`(1-(max(fractions)-min(fractions)))*100`, 

I personally prefer to `std` because `std` is a more formal formula and take account of the fractions on all nodes.

Reply:
I am guessing this was the "Range" option in that discussion, std sounds good.

Reply:
thanks!